phpseclib/File/X509.php

```diff
@@ -3277,7 +3277,17 @@ function sign($issuer, $subject, $signatureAlgorithm = 'sha1WithRSAEncryption')
 
             $startDate = !empty($this->startDate) ? $this->startDate : @date('D, d M Y H:i:s O');
             $endDate = !empty($this->endDate) ? $this->endDate : @date('D, d M Y H:i:s O', strtotime('+1 year'));
-            $serialNumber = !empty($this->serialNumber) ? $this->serialNumber : new Math_BigInteger();
+            if (!empty($this->serialNumber)) {
+                $serialNumber = $this->serialNumber;
+            } else {
+                if (!function_exists('crypt_random_string')) {
+                    include_once 'Crypt/Random.php';
+                }
+                // "The serial number MUST be a positive integer"
+                // "Conforming CAs MUST NOT use serialNumber values longer than 20 octets."
+                //  -- https://tools.ietf.org/html/rfc5280#section-4.1.2.2
+                $serialNumber = new Math_BigInteger(crypt_random_string(20) & ("\x7F" . str_repeat("\xFF", 19)), 256);
```

Comment:
Might be a good idea to explain (in a comment) what this statement does, especially the & part.